src/engine/BMDie.php

```diff
@@ -290,7 +290,13 @@ public function roll($isTriggeredByAttack = FALSE, $isSubdie = FALSE) {
 
         if (!isset($this->value) ||
             ($this->doesReroll && !$this->has_flag('JustPerformedTripAttack'))) {
-            $this->set__value(bm_rand($this->min, $this->max));
+            $hookResultArray = $this->run_hooks(__FUNCTION__, array('die' => $this));
+
+            // if all hook results are FALSE, then roll the die
+            if (empty($hookResultArray) ||
+                (0 == count(array_filter($hookResultArray, function ($value) { return $value !== FALSE; })))) {
+                $this->set__value(bm_rand($this->min, $this->max));
+            }
```

Comment:
Within the `roll()` die hook?  I'm staring at the logic of your diff and just not seeing what you say here --- what i see is that on lines 33-34, `roll()` returns FALSE *if* the die value is already set.  So where does it return FALSE *and* set the die value?  Enh, there's no point arguing about it --- i'm provisionally willing to believe that you're right and i'm just blind here, and i'll test it and find out.